test/tic_tac_toe/session_test.clj

```diff
@@ -0,0 +1,26 @@
+(ns tic-tac-toe.session-test
+  (:require [clojure.test :refer [deftest testing is]]
+            [tic-tac-toe.board :refer :all]
+            [tic-tac-toe.session :refer :all]))
+
+(deftest session-test
+  (testing "validates input from player correctly"
+    (is (= (validate-integer 1) true)))
+  
+  (testing "validates input from player correctly"
+    (is (= (validate-integer 0) false)))
+  
+  (testing "validates input from player correctly"
+    (is (= (validate-integer -1) false)))
+  
+  (testing "validates input from player correctly"
+    (is (= (validate-integer "a") false)))
+  
+  (testing "checks to see if position is already taken"
+    (is (= (empty-space? ["X" 2 3 4 5 6 7 8 9] 1) false)))
+  
+  (testing "prompts player for a valid move at the beginning of the turn"
+    (is (with-in-str "1" (session-turn [1 2 3 4 5 6 7 8 9] "X")) ["X" 2 3 4 5 6 7 8 9]))
+
+  (testing "prompts player for a valid move if invalid move is provided"
+    (is (with-in-str "p\n1" (session-turn [1 2 3 4 5 6 7 8 9] "X")) ["X" 2 3 4 5 6 7 8 9])))
```

Comment:
where are the tests for session-run? It doesn't seem like you test drove these two functions. I would advise that you try writing them over again by test driving them.

Reply:
Yeah I only knew how to do basic unit testing. In order to meet the requirements the stories in a timely manner, I created the session-run as soon as I could instead of learning how to make more complex tests. After the discussion during the last IPM, I should have asked for more time and make the product slowly but properly.